wordpress-mobile/WordPress-Android
# Adding tracking to recommend the app feature.

Refs: pctCYC-8X-p2#comment-340 and  iOS [counter part PR](https://github.com/wordpress-mobile/WordPress-iOS/pull/17027)

Part of #15269 this PR does the following:

- Adds tracking for  `recommend_app_engaged`: this event is fired when the Share Sheet is succesfully opened: it has a `source` parameter that currently is only triggered for "me" given the "about" source is not available
- Adds tracking for `recommend_app_content_fetch_failed`: this event is fired on failure in getting the remote template; it has a `source` parameter (only "me" available) and an `error` parameter that reports the error message. Currently the failure is triggered for
   - No network available
   - Volley network error
   - getTemplateFromJson parsing error
- A bit of refactor of calls provider to ease testability and moving the No network check inside the net calls provider so to have all failure causes in one place inside that class; also it seemed to me a good place where to check for network in the end.

### To test
- Activate the feature flag and the tracking in the app settings and look in logcat for events
- Check the following is triggered on success when opening the sheet (both when network call happens and in successive taps on the share button without leaving the Me screen)

```
🔵 Tracked: recommend_app_engaged, Properties: {"source":"me"}
```

- Check the`recommend_app_content_fetch_failed` event is triggered on following failures:
   - No Network available: setting the airplane mode you should get
```
🔵 Tracked: recommend_app_content_fetch_failed, Properties: {"source":"me","error":"getRecommendTemplate > No Network available"}
```
   - Volley network error, getTemplateFromJson parsing error: forcing conditions with debugger or by charles proxy (error property will depend on how the error is forced)
- Since a bit of refactoring was made, smoke test the feature.

## Regression Notes
1. Potential unintended areas of impact
None.

2. What I did to test those areas of impact (or what existing automated tests I relied on)
Manual testing and unit testing added.

3. What automated tests I added (or what prevented me from doing so)
Unit testing for tracking added.

PR submission checklist:
